src/quirks.rs

```diff
@@ -110,14 +112,23 @@ pub fn set_host(url: &mut Url, new_host: &str) -> Result<(), ()> {
     let opt_port;
     {
         let scheme = url.scheme();
-        let result = Parser::parse_host(Input::new(new_host), SchemeType::from(scheme));
+        let scheme_type = SchemeType::from(scheme);
+        let result = if scheme_type == SchemeType::File {
+            Parser::get_file_host(input)
+        } else {
+            Parser::parse_host(input, scheme_type)
```

Comment:
Why not do what `get_file_host` does directly in `parse_host`, checking in that latter function directly whether `scheme_type` is `SchemeType::File`?